### PR TITLE
Do not fail the build on invalid declared file inputs

### PR DESCRIPTION
**Pull request by wolfs**
Since Gradle 7.0 we evaluate all file inputs of a task, regardless if
we snapshot the inputs. This causes some tasks without output files
to fail now, most prominently the Android Plugin Lint task for
AGP versions < 4.2.

We now catch possible exceptions when evaluating input file properties
and emit a validation warning instead of failing the build. We
can then fail the build for those input file properties in Gradle 8.0.

### CODE REVIEW

**Comment by wolfs** [comment]
This PR also sets the supported AGP version of Gradle 7.0 to AGP 4.0.

**Comment by wolfs** [comment]
@bot-gradle test this

**Comment by wolfs** [comment]
@bot-gradle test this

File: subprojects/docs/src/docs/userguide/compatibility.adoc [hunk 1]
```diff
@@ -30,4 +30,4 @@ Gradle is tested with Kotlin 1.3.21 through 1.4.20.
 Gradle is tested with Groovy 1.5.8 through 2.5.12.
 
 == Android
-Gradle is tested with Android Gradle Plugin 3.4, 3.5, 3.6, 4.0, 4.1 and 4.2. Alpha and beta versions may or may not work.
+Gradle is tested with Android Gradle Plugin 4.0, 4.1, 4.2 and 7.0. Alpha and beta versions may or may not work.
```
Comment (lptr) [inline subprojects/docs/src/docs/userguide/compatibility.adoc:33]: Do we already run the tests for 7.0?

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java [hunk 2]
```diff
@@ -84,31 +85,45 @@ private void detectMissingDependencies(LocalTaskNode node, TypeValidationContext
...
+                            }
+                        }
+
+                        @Override
+                        public void visitFileTreeBackedByFile(File file, FileTreeInternal fileTree, FileSystemMirroringFileTree sourceTree) {
+                            taskInputs.add(file.getAbsolutePath());
+                        }
+                    });
+                } catch (Exception e) {
+                    if (historyMaintained) { // We would try to snapshots the inputs anyway, no need to suppress the exception
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java:116]: Nit: let's not use trailing comments.

**Review by lptr — COMMENTED** [review]
I left a few comments.

File: subprojects/docs/src/docs/userguide/compatibility.adoc (hunk 1 above)
Comment (wolfs) [reply subprojects/docs/src/docs/userguide/compatibility.adoc:33]: Yes.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy [hunk 3]
```diff
@@ -282,6 +282,37 @@ class MissingTaskDependenciesIntegrationTest extends AbstractIntegrationSpec {
...
+            task "broken" {
+                inputs.files(5).withPropertyName("invalidInputFileCollection")
+
+                doLast {
+                    println "success"
+                }
+            }
+        """
+        when:
+        executer.expectDocumentedDeprecationWarning("Consider using Task.dependsOn instead of an input file collection. This behaviour has been deprecated and is scheduled to be removed in Gradle 7.0. Execution optimizations are disabled due to the failed validation. See https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:up_to_date_checks for more details.")
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy:296]: We are not missing the explanation part. This is a multiline deprecation warning and our deprecation infrastructure only picks up the line containing "is deprecated...". The `Property 'invalidInputFileCollection' cannot be resolved:...` is the start of the deprecation warning (+ the regular validation warning deprecation preamble).

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy [hunk 4]
```diff
@@ -282,6 +282,37 @@ class MissingTaskDependenciesIntegrationTest extends AbstractIntegrationSpec {
...
+        run "broken"
+        then:
+        executedAndNotSkipped ":broken"
+        outputContains("""
+            Validation failed for task ':broken', disabling optimizations:
```
Comment (ldaley) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy:301]: @lptr  Does this invalidate our conversation yesterday about 7.0 having no “disabled optimizations”? 

/cc @pioterj @facewindu

**Comment by wolfs** [comment]
@bot-gradle test this

**Review by lptr — APPROVED** [review]
LGTM.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy (hunk 4 above)
Comment (lptr) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy:301]: Let's discuss this next week!

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java [hunk 5]
```diff
@@ -84,31 +85,46 @@ private void detectMissingDependencies(LocalTaskNode node, TypeValidationContext
...
+                        }
+
+                        @Override
+                        public void visitFileTreeBackedByFile(File file, FileTreeInternal fileTree, FileSystemMirroringFileTree sourceTree) {
+                            taskInputs.add(file.getAbsolutePath());
+                        }
+                    });
+                } catch (Exception e) {
+                    if (historyMaintained) {
+                        // We would try to snapshots the inputs anyway, no need to suppress the exception
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java:117]:
```suggestion
                        // We would later try to snapshot the inputs anyway, no need to suppress the exception
```

**Pull request merged**